document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoCollection.java

```diff
@@ -333,6 +334,76 @@ public BulkUpdateResult bulkUpdateSubDocs(Map<Key, Map<String, Document>> docume
     return new BulkUpdateResult(writeResult.getModifiedCount());
   }
 
+  @Override
+  public BulkUpdateResult bulkOperationOnArrayValue(BulkArrayValueUpdateRequest request)
+      throws Exception {
+    List<BasicDBObject> basicDBObjects = new ArrayList<>();
+    try {
+      for (Document subDocument : request.getSubDocuments()) {
+        String jsonString = subDocument.toJson();
+        JsonNode jsonNode = MAPPER.readTree(jsonString);
+        JsonNode sanitizedJsonNode = recursiveClone(jsonNode, this::encodeKey);
+        String sanitizedJsonString = MAPPER.writeValueAsString(sanitizedJsonNode);
+        basicDBObjects.add(BasicDBObject.parse(sanitizedJsonString));
+      }
+    } catch (Exception e) {
+      LOGGER.error(
+          "Exception updating document. keys: {} operation {} subDocPath {} subDocuments :{}",
+          request.getKeys(),
+          request.getOperation(),
+          request.getSubDocPath(),
+          request.getSubDocuments());
+      throw e;
+    }
+    BasicDBObject operationObject;
+    switch (request.getOperation()) {
+      case ADD:
+        operationObject = getAddOperationObject(request.getSubDocPath(), basicDBObjects);
+        break;
+      case REMOVE:
+        operationObject = getRemoveOperationObject(request.getSubDocPath(), basicDBObjects);
+        break;
+      case SET:
+        operationObject = getSetOperationObject(request.getSubDocPath(), basicDBObjects);
+        break;
+      default:
+        throw new UnsupportedOperationException("Unknown operation : " + request.getOperation());
+    }
+
+    List<UpdateManyModel<BasicDBObject>> bulkWriteUpdate =
+        List.of(
+            new UpdateManyModel(
+                selectionCriteriaForKeys(request.getKeys()), operationObject, new UpdateOptions()));
+    BulkWriteResult writeResult = collection.bulkWrite(bulkWriteUpdate);
```

Comment:
So we had talked separately about introducing a new upsert api that supports transactionality. Would it make sense to start that here? My concern is that every new upsert use case just gets its own little individual api, then we start getting combinations and the api keeps growing in a crazy way. I'm basically imagining one flexible api that encompasses all of our upsert needs - accepting one or more operations, with described transactionality, merging, sub docs, etc. That API doesn't need to be built out in a day, but if we could start building new use cases like this one into it, that would be awesome.